File: BinarySearch2/problem1.py
#Time Complexity: O(logn)
#Space Complexity: O(1)
#Did this code successfully run on leetcode: yes
#Any problem you faced while coding thing:
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        #Main idea: seperate the two. Find target such that previous is smaller/next is bigger. normal binary search.
        #if it's target but previous not smaller,move right index to mid, for end position, moveleft index to mid.
        def start_position(nums,target):
            #edge case:
            if not nums:
                return -1
            low=0
            high=len(nums)-1
            while high>=low:
                mid=int(low+ (high-low)/2) #to avoid overflow
                if nums[mid]==target:
                    if mid==0 or nums[mid-1]<target:
                        return mid
                    else:
                        high=mid-1
                elif nums[mid]>target:
                    high=mid-1
                else:
                    low=mid+1
            return -1
        def end_position(nums,target):
            if not nums:
                return -1
            low=0
            high=len(nums)-1
             #to avoid overflow
            while high>=low:
                mid=int(low+ (high-low)/2)
                if nums[mid]==target:
                    if mid==len(nums)-1 or nums[mid+1]>target:
                        return mid
                    else:
                        low=mid+1
                elif nums[mid]>target:
                    high=mid-1
                else:
                    low=mid+1
            return -1

        return [start_position(nums,target),end_position(nums,target)]

File: src/BinarySearch/Problem1.java
package BinarySearch;

public class Problem1 {
	public static int[] searchRange(int[] nums, int target) {
        int left=extremeinsertionindex(nums,target,true);
        int right=extremeinsertionindex(nums,target,false)-1;
        
        int[] res= new int[]{-1,-1};
        if(left==nums.length || nums[left]!=target)
            return res;
        res[0]=left;
        res[1]=right;
        
        return res;
    }
    
    public static int extremeinsertionindex(int[] nums,int target,boolean left){
        int lo=0;
        int hi=nums.length;
        
        while(lo<hi){
            int mid= (lo+hi)/2;
            if(nums[mid]>target || (left && nums[mid]==target)){
                hi=mid;
            }else{
                lo=mid+1;
            }
        }
        
        return lo;
    }
    
    public static void main(String args[]) {
    	int[] nums=new int[] {2,3,4,5,6,6,6,7,7,7,8,9};
    	int target=6;
    	int[] res= searchRange(nums,target);
    	
    	System.out.println(res[0]);
    	System.out.println(res[1]);
    }
}
